lib/ZumoOta/src/BootloaderCom.cpp

```diff
@@ -97,26 +102,122 @@ class IntProgCmdProvider: public CmdProvider {
+ *
+ * This class provides commands for firmware programming, specifically tailored
+ * for the Zumo32U4 device.
+ */
 class FwProgCmdProvider : public CmdProvider {
     public:
-        FwProgCmdProvider(FileHandle * fs);
-}
-#endif
+        /**
+         * @brief Constructor for the FwProgCmdProvider class.
+         * @param fileName The name of the file to be used for firmware programming.
+         * Initializes a new instance of the FwProgCmdProvider class with the provided file name.
+         * @remarks The file specified by fileName should contain the firmware to be programmed.
+         */
+         FwProgCmdProvider(const std::string& fileName):
+            m_index(0),
+            m_fileName(fileName)
+    {}
+
+    /**
+     * @brief Resets the command provider to its initial state.
+     *
+     * This method resets the internal index that tracks the current command
+     * in the command sequence memory.
+     */
+    virtual void reset() { m_index = 0;}
+
+    /**
+     * @brief Retrieves the next command and the next response.
+     *
+     * This method returns the next command and its associated response
+     * from the command sequence memory.
+     *
+     * @param cmd Reference to a pointer to the next command.
+     * @param rsp Reference to a pointer to the next response.
+     * @return True if there is a next command and next response, false otherwise.
+     */
+    virtual bool next( CommandInfo *& cmd, ResponseInfo *& rsp)
+    {
+        /*Determine the command and response for the current chunk.*/
+        uint32_t commandIndex = m_index % SEQ_LENGHT;
+        cmd = &m_cmds[commandIndex];
+
+        if(cmd->command == Zumo32U4Specification::SET_MEMORY_ADDR)
+        {
+            cmd->commandsize = FinalLength;
+        }
+        else if(cmd->command == Zumo32U4Specification::WRITE_MEMORY_PAGE)
+        {
+            size_t bytesRead = m_fileManager.read128Bytes(m_fileName.c_str(), m_buffer);
+            /*If no bytes were read, there are no more chunks left.*/
+            if (0 == bytesRead)
+            {
+                cmd = nullptr;
+                rsp = nullptr;
+                return false;
+            }
+            else
+            {
+                cmd->commandsize += bytesRead;
+            }
+        }
+        rsp = &m_responses[commandIndex];
+
+        /* Move to the next chunk.*/
+        ++m_index;
+
+        return true;
+    }
+
+    FileManager m_fileManager; /**< Instance of Filemanger class.*/
+    const size_t FinalLength = 3;/**< Final Size of the updated Command.*/
+
+private:
+    uint8_t m_index; /**< current index in sequence.*/
+    static  CommandInfo m_cmds[]; /**< Array of command information.*/
+    static  ResponseInfo m_responses[]; /**< Array of response information.*/
+    std::string m_fileName = "/zumo_firmware.bin"; /**< Name of the file to be read.*/
+    static const uint32_t SEQ_LENGHT = 2U; /**< Length of the command sequence.*/

```

Comment:
I don't understand what length this is. The programming is for sure longer then 2 commands.

It seems just to be a toggle between set_addr and writ_page.  So why not be explicit with somthing like 

enum ProgCmds  {FLSPRG_SET_ADDR, FLSPRG_WRITE_BLOCK};
ProgrCmds nextCmd = FLSPRG_SET_ADDR;

instead of the misleading SEQ_LENGTH. This would be way better to understand.

Reply:
It will be adjusted in the next commit.